File: daemon/doc2pdf.py
# IRIS-Face · © 2026 Sejun Ham (함세준) · MIT · https://feynman520.github.io/card/#home
# -*- coding: utf-8 -*-
"""
doc2pdf.py — 문서(한/글·워드·엑셀·파워포인트)를 PDF로 변환한다.

IRIS-Face 데몬이 부르는 독립 실행 스크립트. 대화 안에 문서를 삽입하려고 원본을 PDF로
한 번 바꿔 캐시에 둔다. 원본은 절대 건드리지 않는다(읽기 전용으로 열고 새 PDF만 만든다).

사용:  python doc2pdf.py <원본경로> <PDF저장경로>
성공:  종료코드 0 + stdout 마지막 줄 "OK <PDF경로>"
실패:  종료코드 1 + stderr 에 사유

인터프리터:
  - .hwp/.hwpx → pyhwpx 가 깔린 한/글 MCP venv 파이썬으로 실행해야 한다(데몬이 골라 준다).
  - .doc(x)/.xls(x)/.ppt(x) → win32com 이 있는 아무 파이썬(시스템 파이썬 가능).
"""
from __future__ import annotations

import os
import sys

# 윈도 cp949 콘솔에서 한글 print 가 죽지 않도록 (R-010)
try:
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
    sys.stderr.reconfigure(encoding="utf-8", errors="replace")
except Exception:
    pass


def _err(msg: str) -> None:
    print(msg, file=sys.stderr, flush=True)


def _ensure_parent(path: str) -> None:
    os.makedirs(os.path.dirname(os.path.abspath(path)), exist_ok=True)


# ---------------------------------------------------------------- 한/글 --------
def _ensure_hwp_security() -> None:
    """한/글 보안승인모듈(FilePathChecker)을 HKCU 레지스트리에 선등록한다.

    없으면 한/글이 파일 입출력 때 팝업을 띄워 무인 변환이 멈춘다. pyhwpx 번들 DLL 을
    찾아 레지스트리 값만 써 둔다(이미 있으면 그대로). MCP 세션 코드와 같은 방식."""
    try:
        import winreg

        import pyhwpx

        dll = os.path.join(os.path.dirname(os.path.abspath(pyhwpx.__file__)),
                           "FilePathCheckerModule.dll")
        if not os.path.exists(dll):
            return
        for sub in (r"Software\HNC\HwpAutomation\Modules",
                    r"Software\Hnc\HwpUserAction\Modules"):
            try:
                key = winreg.CreateKeyEx(winreg.HKEY_CURRENT_USER, sub, 0, winreg.KEY_WRITE)
                winreg.SetValueEx(key, "FilePathCheckerModule", 0, winreg.REG_SZ, dll)
                winreg.CloseKey(key)
            except OSError:
                continue
    except Exception as e:  # 등록 실패해도 변환은 시도한다
        _err(f"[hwp] 보안모듈 등록 건너뜀: {e!r}")


def hwp_to_pdf(src: str, out: str) -> None:
    _ensure_hwp_security()
    from pyhwpx import Hwp

    hwp = Hwp(new=True, visible=False)
    try:
        if not hwp.open(src):
            raise RuntimeError(f"한/글 열기 실패: {src}")
        ok = hwp.save_as(out)  # 확장자 .pdf → 내부 PDF 액션
        if not ok or not os.path.exists(out):
            raise RuntimeError(f"PDF 저장 실패(반환={ok})")
    finally:
        try:
            hwp.clear()
        except Exception:
            pass
        try:
            hwp.quit()
        except Exception:
            pass


# --------------------------------------------------------------- 오피스 --------
def _office_to_pdf(src: str, out: str, ext: str) -> None:
    import pythoncom
    import win32com.client as win32

    pythoncom.CoInitialize()
    try:
        if ext in (".doc", ".docx", ".rtf", ".odt"):
            app = win32.DispatchEx("Word.Application")
            app.Visible = False
            try:
                doc = app.Documents.Open(src, ReadOnly=True)
                doc.ExportAsFixedFormat(OutputFileName=out, ExportFormat=17)  # wdExportFormatPDF
                doc.Close(False)
            finally:
                app.Quit()
        elif ext in (".xls", ".xlsx", ".xlsm", ".ods"):
            app = win32.DispatchEx("Excel.Application")
            app.Visible = False
            app.DisplayAlerts = False
            try:
                wb = app.Workbooks.Open(src, ReadOnly=True)
                wb.ExportAsFixedFormat(0, out)  # 0 = xlTypePDF, 시트 전체
                wb.Close(False)
            finally:
                app.Quit()
        elif ext in (".ppt", ".pptx", ".odp"):
            app = win32.DispatchEx("PowerPoint.Application")
            try:
                pres = app.Presentations.Open(src, ReadOnly=True, WithWindow=False)
                pres.SaveAs(out, 32)  # 32 = ppSaveAsPDF
                pres.Close()
            finally:
                app.Quit()
        else:
            raise RuntimeError(f"지원하지 않는 오피스 형식: {ext}")
    finally:
        try:
            pythoncom.CoUninitialize()
        except Exception:
            pass
    if not os.path.exists(out):
        raise RuntimeError("PDF가 생성되지 않았습니다.")


# ----------------------------------------------------------------- main --------
def main() -> int:
    if len(sys.argv) != 3:
        _err("사용법: python doc2pdf.py <원본> <PDF저장경로>")
        return 2
    src, out = os.path.abspath(sys.argv[1]), os.path.abspath(sys.argv[2])
    if not os.path.exists(src):
        _err(f"원본을 찾을 수 없습니다: {src}")
        return 1
    _ensure_parent(out)
    ext = os.path.splitext(src)[1].lower()
    try:
        if ext in (".hwp", ".hwpx"):
            hwp_to_pdf(src, out)
        else:
            _office_to_pdf(src, out, ext)
    except Exception as e:
        _err(f"변환 실패({ext}): {e!r}")
        return 1
    print(f"OK {out}", flush=True)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: daemon/doc2pdf.mjs
// IRIS-Face · © 2026 Sejun Ham (함세준) · MIT · https://feynman520.github.io/card/#home
// doc2pdf.mjs — 문서(한/글·워드·엑셀·PPT)를 PDF로 변환해 캐시에 두고 그 경로를 돌려준다.
// 데몬은 순수 node 라 COM 을 직접 못 부른다 → 파이썬(doc2pdf.py)을 자식으로 띄운다.
//   · .hwp/.hwpx → pyhwpx 가 깔린 한/글 MCP venv 파이썬 (환경변수 IRIS_FACE_HWP_PY 로 덮어쓰기)
//   · 그 밖(오피스) → win32com 이 있는 시스템 파이썬 (환경변수 IRIS_FACE_PY / python)
// 변환은 원본을 읽기 전용으로만 열고 새 PDF 를 만든다. 실패하면 throw → 호출부가 카드로 대체.
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import { spawn } from 'node:child_process';
import { fileURLToPath } from 'node:url';
import { toolPath } from './paths.mjs';

const HERE = path.dirname(fileURLToPath(import.meta.url));
const PY = path.join(HERE, 'doc2pdf.py');

export const DOC_EXTS = ['.hwp', '.hwpx', '.doc', '.docx', '.rtf', '.odt', '.xls', '.xlsx', '.xlsm', '.ods', '.ppt', '.pptx', '.odp'];

// 한/글은 pyhwpx 가 필요하다. 환경변수(IRIS_FACE_HWP_PY, PC마다 다른 실제 자리)를 우선 쓰고,
// 없으면 공유 도구 폴더(_agent\shared\tools\document-mcp\hwp\<커밋>\) 아래를 뒤져 venv 를 찾는다.
// 못 찾으면 hwp 변환만 불가(오피스·PDF 는 계속 됨).
const OFFICE_PY = process.env.IRIS_FACE_PY || 'python';

function hwpPyCandidatesFromSharedTools() {
  const base = toolPath('document-mcp', 'hwp');
  const found = [];
  try {
    for (const sub of fs.readdirSync(base, { withFileTypes: true })) {
      if (!sub.isDirectory()) continue;
      const subDir = path.join(base, sub.name);
      found.push(path.join(subDir, '.venv', 'Scripts', 'python.exe'));
      found.push(path.join(subDir, 'hwp-form-automation-mcp', '.venv', 'Scripts', 'python.exe'));
    }
  } catch {}
  return found;
}

function hwpPython() {
  const candidates = [process.env.IRIS_FACE_HWP_PY, ...hwpPyCandidatesFromSharedTools()].filter(Boolean);
  for (const c of candidates) { try { if (fs.existsSync(c)) return c; } catch {} }
  return null;
}

export function isConvertible(file) {
  return DOC_EXTS.includes(path.extname(file).toLowerCase());
}

// 변환 결과 PDF 의 캐시 경로 = 원본 경로 + 수정시각 지문. 원본이 바뀌면 새로 변환한다.
function cachePath(cacheDir, src) {
  const st = fs.statSync(src);
  const key = crypto.createHash('sha1').update(`${src}|${st.mtimeMs}|${st.size}`).digest('hex').slice(0, 16);
  const base = path.basename(src).replace(/[\\/:*?"<>|]/g, '_').slice(0, 60);
  return path.join(cacheDir, `${key}-${base}.pdf`);
}

const inflight = new Map(); // 같은 파일 동시 요청은 한 번만 변환

// src(문서) → PDF 경로. 캐시 적중 시 즉시 반환. cacheDir = state/doc-cache.
export function toPdf(src, cacheDir, log = () => {}) {
  const ext = path.extname(src).toLowerCase();
  if (!fs.existsSync(src)) return Promise.reject(new Error(`파일이 없습니다: ${src}`));
  if (!isConvertible(src)) return Promise.reject(new Error(`변환할 수 없는 형식: ${ext}`));
  fs.mkdirSync(cacheDir, { recursive: true });
  const out = cachePath(cacheDir, src);
  if (fs.existsSync(out) && fs.statSync(out).size > 0) return Promise.resolve(out);
  if (inflight.has(out)) return inflight.get(out);

  const isHwp = ext === '.hwp' || ext === '.hwpx';
  const py = isHwp ? hwpPython() : OFFICE_PY;
  if (isHwp && !py) return Promise.reject(new Error('한/글 변환기(pyhwpx)를 찾지 못했습니다. 한/글이 설치된 PC에서만 미리보기가 됩니다.'));

  const t0 = Date.now();
  const job = new Promise((resolve, reject) => {
    let stderr = '';
    // PYTHONUTF8: 경로에 든 〖 같은 글자를 win32com makepy 가 cp949 콘솔로 print 하다 죽는 것을 막는다(R-010 계열).
    const child = spawn(py, [PY, src, out], { windowsHide: true, env: { ...process.env, PYTHONUTF8: '1', PYTHONIOENCODING: 'utf-8' } });
    const timer = setTimeout(() => { try { child.kill(); } catch {} reject(new Error('변환 시간 초과(90초)')); }, 90_000);
    child.stderr.on('data', (d) => { stderr += d.toString(); });
    child.on('error', (e) => { clearTimeout(timer); reject(e); });
    child.on('close', (code) => {
      clearTimeout(timer);
      if (code === 0 && fs.existsSync(out) && fs.statSync(out).size > 0) {
        log(`doc2pdf ${ext} ${((Date.now() - t0) / 1000).toFixed(1)}s ${path.basename(src)}`);
        resolve(out);
      } else {
        reject(new Error(stderr.trim().split('\n').pop() || `변환 실패(code=${code})`));
      }
    });
  }).finally(() => inflight.delete(out));
  inflight.set(out, job);
  return job;
}
